impeller/entity/geometry/round_superellipse_geometry.cc

```diff
@@ -0,0 +1,306 @@
+  // https://math.stackexchange.com/questions/2573746/superellipse-parametric-equation
+  {
+    const Scalar target_slope = yJ / xJ;
+    Scalar angle = 0;
+    while (true) {
+      const Scalar x = a * pow(abs(sinf(angle)), 2 / n);
+      const Scalar y = a * pow(abs(cosf(angle)), 2 / n);
+      if (y <= target_slope * x) {
+        break;
+      }
+      points.emplace_back(x + s, y + s);
+      angle += kAngleStep;
+    }
+  }
+  // Circular arc JM (B inclusive, M exclusive)
+  DrawCircularArc(points, {xJ + s, yJ + s}, pointM, R);
+
+  // Apply `flip` and `center`.
+  if (!flip) {
+    for (const Point& point : points) {
+      output.push_back(point + center);
+    }
+  } else {
+    for (size_t i = 0; i < points.size(); i++) {
+      const Point& point = points[points.size() - i - 1];
+      output.emplace_back(point.y + center.x, point.x + center.y);
+    }
+  }
+}
+
+static Scalar LimitRadius(Scalar corner_radius, const Rect& bounds) {
+  return std::min(corner_radius,
+                  std::min(bounds.GetWidth() / 2, bounds.GetHeight() / 2));
+}
+
+RoundSuperellipseGeometry::RoundSuperellipseGeometry(const Rect& bounds,
+                                                     Scalar corner_radius)
+    : bounds_(bounds), corner_radius_(LimitRadius(corner_radius, bounds)) {}
+
+RoundSuperellipseGeometry::~RoundSuperellipseGeometry() {}
+
+GeometryResult RoundSuperellipseGeometry::GetPositionBuffer(
+    const ContentContext& renderer,
+    const Entity& entity,
+    RenderPass& pass) const {
+  const Size size = bounds_.GetSize();
+  const Point center = bounds_.GetCenter();
+
+  // The full shape is divided into 4 segments: the top and bottom edges come
+  // from two square-like rounded superellipses (width-aligned), while the left
+  // and right squircles come from another two (height-aligned).
+  //
+  // Denote the distance from the center of the square-like squircles to the
+  // origin as `c`. The width-aligned square-like squircle and the
+  // height-aligned one have the same offset in different directions.
+  const Scalar c = (size.width - size.height) / 2;
+
+  // Draw the first quadrant of the shape and store in `points`. It will be
+  // mirrored to other quadrants later.
+  std::vector<Point> points;
+  points.reserve(41);
+
+  DrawOctantSquareLikeSquircle(points, size.width, corner_radius_, Point{0, -c},
+                               false);
+  points.push_back(Point(size / 2) - gap(corner_radius_));  // Point M
+  DrawOctantSquareLikeSquircle(points, size.height, corner_radius_, Point{c, 0},
+                               true);
+
+  static constexpr Point reflection[4] = {{1, 1}, {1, -1}, {-1, -1}, {-1, 1}};
+
+  // Reflect the 1/4 arc into the 4 quadrants and generate the tessellated mesh.
+  // The iteration order is reversed so that the trianges are continuous from
+  // quadrant to quadrant.
+  std::vector<Point> geometry;
+  geometry.reserve(1 + 4 * points.size());
```

Comment:
Once you know the amount of storage needed, its better to write directly into the host buffer as it avoids an extra copy.

Example:

```

 BufferView point_buffer = host_buffer.Emplace(
      nullptr, sizeof(Point) * point_count, alignof(Point));
Point* point_data =  reinterpret_cast<Point*>(point_buffer.GetBuffer()->OnGetContents() +
                                   point_buffer.GetRange().offset);
/// write to point_data                                   
 ```